executor/executor_darwin.h

```diff
@@ -0,0 +1,118 @@
+
+	return __syscall(c->sys_nr, a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7], a[8]);
+}
+
+static void cover_open(cover_t* cov, bool extra)
+{
+	int fd = ksancov_open();
+	if (fd == -1)
+		fail("open of /dev/ksancov failed");
+	if (dup2(fd, cov->fd) < 0)
+		failmsg("failed to dup cover fd", "from=%d, to=%d", fd, cov->fd);
+	close(fd);
+
+	// Note: In the other KCOV implementations we pass the shared memory size
+	// to the initial ioctl, before mmaping. KSANCOV reversed this logic.
+	// Here we instead pass the maximum number of traced PCs to the initial
+	// KSANCOV_IOC_TRACE ioctl. We then pass a size_t pointer to the second
+	// KSANCOV_IOC_MAP ioctl, hence the kernel is instead telling us the final
+	// size. We have a sanity check in executor.cc checking that cov.size isn't
+	// larger or equal to kCoverSize. To make sure that assumption holds, we're
+	// calculating the max_entries accordingly.
+	size_t max_entries = floor(
+	    (kCoverSize - sizeof(struct ksancov_trace)) / sizeof(uint32_t));
+
+	// Note: XNUs KSANCOV API forces us to choose the mode after opening the
+	// device and before mmaping the coverage buffer. As the function we are
+	// in, cover_open(), expects us to mmap here, we are forced to commit to a
+	// mode here as well. For other OSes we commit to a mode in cover_enable(),
+	// based on collect_comps. This is not really a problem though, as TRACE_PC
+	// is the only relevant mode for us for now. XNU doesn't support TRACE_CMP
+	// and we don't care about the counters/nedges modes in XNU.
+	if (ksancov_mode_trace(cov->fd, max_entries))
+		fail("ioctl init trace write failed");
+
+	uintptr_t mmap_ptr = 0;
+	size_t mmap_alloc_size = 0;
+	if (ksancov_map(cov->fd, &mmap_ptr, &mmap_alloc_size))
+		fail("cover mmap failed");
+
+	// Sanity check to make sure our assumptions in the max_entries calculation
+	// hold up.
+	if (mmap_alloc_size > kCoverSize)
+		fail("mmap allocation size larger than anticipated");
+
+	cov->data = (char*)mmap_ptr;
+	cov->data_end = cov->data + mmap_alloc_size;
+}
+
+static void cover_protect(cover_t* cov)
+{
+}
+
+static void cover_unprotect(cover_t* cov)
+{
+}
+
+static void cover_enable(cover_t* cov, bool collect_comps, bool extra)
+{
+	if (collect_comps)
+		fail("TRACE_CMP not implemented on darwin");
+	if (extra)
+		fail("Extra coverage collection not implemented on darwin");
+	// Note: we are already comitted to TRACE_PC here, hence we don't make use
+	// of collect_comps. For more details see the comment in cover_open().
+	if (ksancov_thread_self(cov->fd))
+		exitf("cover enable write trace failed");
+}
+
+static void cover_reset(cover_t* cov)
+{
+	ksancov_reset((struct ksancov_header*)cov->data);
+	ksancov_start((struct ksancov_header*)cov->data);
+}
+
+static void cover_collect(cover_t* cov)
+{
+	struct ksancov_trace* trace = (struct ksancov_trace*)cov->data;
+	cov->size = ksancov_trace_head(trace);
+	cov->data_offset = ((int64_t) & (trace->pcs)) - ((int64_t)(cov->data));
```

Comment:
data_offset is measured in elements of coverage. Here you calculate it in bytes. I think we need to divide this by sizeof(uint32).

Reply:
I acknowledge what you are saying is correct, but it felt more natural to me, to keep this offset byte based. Particularly because I imagine that offset between cov->data and trace->pcs is likely not evenly divisible by sizeof(uint32).

Hence I instead fixed the usage of cov->data_offset instead.